crypto/evp/evp_lib.c

```diff
@@ -637,14 +647,24 @@ int EVP_CIPHER_get_key_length(const EVP_CIPHER *cipher)
 
 int EVP_CIPHER_CTX_get_key_length(const EVP_CIPHER_CTX *ctx)
 {
-    int ok;
-    size_t v = ctx->key_len;
-    OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
-
-    params[0] = OSSL_PARAM_construct_size_t(OSSL_CIPHER_PARAM_KEYLEN, &v);
-    ok = evp_do_ciph_ctx_getparams(ctx->cipher, ctx->algctx, params);
-
-    return ok != 0 ? (int)v : EVP_CTRL_RET_UNSUPPORTED;
+    if (ctx->key_len < 0) {
+        /*-
+         * Casting away the const is annoying but required here.  We need to
+         * cache the result for performance reasons.
+         */
+        int *const len = &((EVP_CIPHER_CTX *)ctx)->key_len;
+        int ok;
+        OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+
+        params[0] = OSSL_PARAM_construct_int(OSSL_CIPHER_PARAM_KEYLEN, len);
+        ok = evp_do_ciph_ctx_getparams(ctx->cipher, ctx->algctx, params);
+        if (ok == EVP_CTRL_RET_UNSUPPORTED)
+            /* legacy algorithm */
+            *len = ctx->cipher->key_len;
```

Comment:
This is incorrect, legacy never did this.  If you look at `EVP_CIPHER_CTX_key_length()` in 1.1.1, it simply returns `ctx->key_len` and never falls back on the cipher key length.

Reply:
Without [duplicating the key_len field](https://github.com/openssl/openssl/pull/17543#discussion_r787485867), how should I do this?

Reply:
Should have this fixed.  Changed things so that the key_len isn't touched for legacy ciphers (except by the set key length code).